In [1]:
import os
import pandas as pd 
import numpy as np
%matplotlib inline 
import matplotlib.pyplot as plt
import seaborn as sns
import ast 
from ast import literal_eval

In [2]:
path = "C:\\Users\\user\\Desktop\\ESAA 프로젝트"
os.chdir(path)

In [3]:
data = pd.read_csv('movies_metadata.csv')

C:\Users\user\.conda\envs\test\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
pd.value_counts(data["adult"])

False                                                                                                                             45454
True                                                                                                                                  9
 - Written by Ørnås                                                                                                                   1
 Rune Balot goes to a casino connected to the October corporation to try to wrap up her case once and for all.                        1
 Avalanche Sharks tells the story of a bikini contest that turns into a horrifying affair when it is hit by a shark avalanche.        1
Name: adult, dtype: int64

In [5]:
data=data[data["adult"]=="False"]

In [6]:
data[data["adult"]!="False"]

Empty DataFrame
Columns: [adult, belongs_to_collection, budget, genres, homepage, id, imdb_id, original_language, original_title, overview, popularity, poster_path, production_companies, production_countries, release_date, revenue, runtime, spoken_languages, status, tagline, title, video, vote_average, vote_count]
Index: []

[0 rows x 24 columns]

In [7]:
pd.value_counts(data["adult"])

False    45454
Name: adult, dtype: int64

## 2. belongs_to_collection
collection이 있는 관측치는 collection name 만 뽑아서 저장

In [8]:
data["belongs_to_collection"].describe()

count                                                  4491
unique                                                 1695
top       {'id': 415931, 'name': 'The Bowery Boys', 'pos...
freq                                                     29
Name: belongs_to_collection, dtype: object

In [9]:
data["belongs_to_collection"].notnull()

0         True
1        False
2         True
3        False
4         True
         ...  
45461    False
45462    False
45463    False
45464    False
45465    False
Name: belongs_to_collection, Length: 45454, dtype: bool

In [10]:
data["belongs_to_collection"]

0        {'id': 10194, 'name': 'Toy Story Collection', ...
1                                                      NaN
2        {'id': 119050, 'name': 'Grumpy Old Men Collect...
3                                                      NaN
4        {'id': 96871, 'name': 'Father of the Bride Col...
                               ...                        
45461                                                  NaN
45462                                                  NaN
45463                                                  NaN
45464                                                  NaN
45465                                                  NaN
Name: belongs_to_collection, Length: 45454, dtype: object

In [11]:
type(data["belongs_to_collection"][0])

str

In [12]:
type(data["belongs_to_collection"][1])

float

In [13]:
collection=[]
for string in data["belongs_to_collection"]:
    if type(string)==float:
        collection.append(float("Nan"))
    else:
        col_dict=eval(string)
        collection.append(col_dict["name"])

In [14]:
data["belongs_to_collection"]=collection

In [15]:
data["belongs_to_collection"]

0                  Toy Story Collection
1                                   NaN
2             Grumpy Old Men Collection
3                                   NaN
4        Father of the Bride Collection
                      ...              
45461                               NaN
45462                               NaN
45463                               NaN
45464                               NaN
45465                               NaN
Name: belongs_to_collection, Length: 45454, dtype: object

## 3. budget
type을 float로 변경

In [16]:
data["budget"].describe()

count     45454
unique     1223
top           0
freq      36565
Name: budget, dtype: object

In [17]:
data["budget"]=pd.to_numeric(data["budget"])

In [18]:
data["budget"].describe()

count    4.545400e+04
mean     4.225399e+06
std      1.742576e+07
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      3.800000e+08
Name: budget, dtype: float64

## 4. genres
genres name만 포함하는 리스트 형태로 저장

In [19]:
data["genres"].describe()

count                             45454
unique                             4066
top       [{'id': 18, 'name': 'Drama'}]
freq                               5000
Name: genres, dtype: object

In [20]:
data["genres"]

0        [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
1        [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...
2        [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...
3        [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...
4                           [{'id': 35, 'name': 'Comedy'}]
                               ...                        
45461    [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...
45462                        [{'id': 18, 'name': 'Drama'}]
45463    [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...
45464                                                   []
45465                                                   []
Name: genres, Length: 45454, dtype: object

In [21]:
data["genres"][0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [22]:
data["genres"][45464]

'[]'

In [23]:
eval(data["genres"][0])

[{'id': 16, 'name': 'Animation'},
 {'id': 35, 'name': 'Comedy'},
 {'id': 10751, 'name': 'Family'}]

In [24]:
genres=[]
for string in data["genres"]:
    genres_list=eval(string)
    if len(genres_list)==0:
        genres.append(float("Nan"))
    else:
        genres2=[]
        for genres_dict in genres_list:
            genres_name=genres_dict["name"]
            genres2.append(genres_name)
        genres.append(genres2)

In [25]:
data["genres"]=genres

In [26]:
data["genres"]

0         [Animation, Comedy, Family]
1        [Adventure, Fantasy, Family]
2                   [Romance, Comedy]
3            [Comedy, Drama, Romance]
4                            [Comedy]
                     ...             
45461                 [Drama, Family]
45462                         [Drama]
45463       [Action, Drama, Thriller]
45464                             NaN
45465                             NaN
Name: genres, Length: 45454, dtype: object

## 5. homepage
변수 삭제

In [27]:
del data["homepage"]

## 6. id
type를 object에서 int로 변경 후 같은 값을 갖는 관측치는 하나만 남겨놓고 삭제

In [28]:
data["id"].describe()

count      45454
unique     45424
top       141971
freq           3
Name: id, dtype: object

In [29]:
data["id"]=pd.to_numeric(data["id"],errors='coerce')

In [30]:
pd.value_counts(data["id"])

141971    3
22649     2
12600     2
109962    2
14788     2
         ..
376934    1
86751     1
19164     1
29403     1
276895    1
Name: id, Length: 45424, dtype: int64

In [31]:
data[data["id"]==141971]

adult belongs_to_collection  budget               genres      id  \
13261  False                   NaN       0  [Thriller, Mystery]  141971   
13375  False                   NaN       0  [Thriller, Mystery]  141971   
16764  False                   NaN       0  [Thriller, Mystery]  141971   

         imdb_id original_language original_title  \
13261  tt1180333                fi       Blackout   
13375  tt1180333                fi       Blackout   
16764  tt1180333                fi       Blackout   

                                                overview popularity  ...  \
13261  Recovering from a nail gun shot to the head an...   0.411949  ...   
13375  Recovering from a nail gun shot to the head an...   0.411949  ...   
16764  Recovering from a nail gun shot to the head an...   0.411949  ...   

      release_date revenue runtime                        spoken_languages  \
13261   2008-12-26     0.0   108.0  [{'iso_639_1': 'fi', 'name': 'suomi'}]   
13375   2008-12-26     0.0   108.0  [{'iso_639_1': 'fi', 'name': 'suomi'}]   
16764   2008-12-26     0.0   108.0  [{'iso_639_1': 'fi', 'name': 'suomi'}]   

         status                                            tagline     title  \
13261  Released  Which one is the first to return - memory or t...  Blackout   
13375  Released  Which one is the first to return - memory or t...  Blackout   
16764  Released  Which one is the first to return - memory or t...  Blackout   

       video vote_average vote_count  
13261  False          6.7        3.0  
13375  False          6.7        3.0  
16764  False          6.7        3.0  

[3 rows x 23 columns]

In [32]:
data=data.drop_duplicates(["id"],keep="first")

In [33]:
pd.value_counts(data["id"])

2047      1
46492     1
38264     1
327029    1
62836     1
         ..
33765     1
176867    1
376934    1
86751     1
276895    1
Name: id, Length: 45424, dtype: int64

In [34]:
np.sum(data.duplicated(["id"]))

0

## 7. imdb_id
제거

In [35]:
data["imdb_id"].describe()

count         45407
unique        45407
top       tt1079980
freq              1
Name: imdb_id, dtype: object

In [36]:
del data["imdb_id"]

In [37]:
data.head(3)

adult      belongs_to_collection    budget                        genres  \
0  False       Toy Story Collection  30000000   [Animation, Comedy, Family]   
1  False                        NaN  65000000  [Adventure, Fantasy, Family]   
2  False  Grumpy Old Men Collection         0             [Romance, Comedy]   

      id original_language    original_title  \
0    862                en         Toy Story   
1   8844                en           Jumanji   
2  15602                en  Grumpier Old Men   

                                            overview popularity  \
0  Led by Woody, Andy's toys live happily in his ...    21.9469   
1  When siblings Judy and Peter discover an encha...    17.0155   
2  A family wedding reignites the ancient feud be...    11.7129   

                        poster_path  ... release_date      revenue runtime  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  ...   1995-10-30  373554033.0    81.0   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg  ...   1995-12-15  262797249.0   104.0   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg  ...   1995-12-22          0.0   101.0   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline             title  video  \
0                                                NaN         Toy Story  False   
1          Roll the dice and unleash the excitement!           Jumanji  False   
2  Still Yelling. Still Fighting. Still Ready for...  Grumpier Old Men  False   

  vote_average vote_count  
0          7.7     5415.0  
1          6.9     2413.0  
2          6.5       92.0  

[3 rows x 22 columns]

## 8. original_language
모든 변수 전처리가 끝난 후에 영어 영화만 따로 모은 데이터 추출 예정
Nan인 것들은 뒤에 spoken_language 변수 정리할 때 다시 정리

In [38]:
data["original_language"].describe()

count     45413
unique       89
top          en
freq      32246
Name: original_language, dtype: object

In [39]:
pd.value_counts(data["original_language"])

en    32246
fr     2436
it     1528
ja     1347
de     1079
      ...  
la        1
ay        1
qu        1
uz        1
lb        1
Name: original_language, Length: 89, dtype: int64

In [40]:
np.sum(pd.isnull(data["original_language"]))

11

In [41]:
data[pd.isnull(data["original_language"])==True]

adult belongs_to_collection  budget                         genres  \
19574  False                   NaN       0                  [Documentary]   
21602  False                   NaN       0               [Romance, Drama]   
22832  False                   NaN       0                 [War, Western]   
32141  False                   NaN       0                       [Comedy]   
37407  False                   NaN       0                    [Animation]   
41047  False                   NaN       0                  [Documentary]   
41872  False                   NaN       0                  [Documentary]   
44057  False                   NaN       0                        [Drama]   
44410  False                   NaN       0       [Animation, Documentary]   
44576  False                   NaN       0                            NaN   
44655  False                   NaN       0  [Action, War, Drama, History]   

           id original_language              original_title  \
19574  283101               NaN     Shadowing the Third Man   
21602  103902               NaN              Unfinished Sky   
22832  359195               NaN             13 Fighting Men   
32141  147050               NaN                   Lambchops   
37407  257095               NaN                      Bajaja   
41047  332742               NaN              Song of Lahore   
41872  144410               NaN  Annabelle Serpentine Dance   
44057  380438               NaN       Lettre d'une inconnue   
44410  381096               NaN                        Garn   
44576  381525               NaN                      WiNWiN   
44655  104473               NaN       La prise de Tournavos   

                                                overview popularity  \
19574  Documentary about the production of The Third ...   0.017007   
21602  An Outback farmer takes in an Afghani woman wh...   0.359818   
22832  A group of Union Army soldiers is charged with...   0.070647   
32141  George and Gracie enter an elegant drawing roo...   0.073418   
37407  The first fairy tale transformed into a full-l...   0.036841   
41047  Until the late 1970s, the Pakistani city of La...   0.373688   
41872  Two dancers perform in succession facing a sta...    1.58434   
44057                                                NaN   0.001283   
44410  The traditional crafts of crochet and knitting...   0.067624   
44576  American investment fund buys Austrian compani...   0.030766   
44655  Three military men, seen inside a fortificatio...   0.471325   

                            poster_path  ... release_date revenue runtime  \
19574  /xT9csxMKYbizRQgvNzCXJ67R7er.jpg  ...   2004-10-11     0.0    95.0   
21602  /vfaJuZF0ocVwBEDYZhoULGuVxrz.jpg  ...   2007-08-04     0.0    94.0   
22832  /q5ADXvvzYtAOyrReIg1HckdPj62.jpg  ...   1960-03-31     0.0    69.0   
32141  /upQByEwY58sHWdL3RZm7kmRc44s.jpg  ...   1929-10-31     0.0     8.0   
37407  /12wf2EkrPGxNzRbNyoau8GIFp3M.jpg  ...   1950-01-26     0.0    87.0   
41047  /6iJqscs2xQLKKvrTCWmlHd9rf3W.jpg  ...   2015-11-13     0.0    82.0   
41872  /l7Lzvah6PL6vdMF5p0elKiyYqp9.jpg  ...   1895-04-01     0.0     1.0   
44057  /gk32ZWTYgfTXctEyCDAWeMcxo7X.jpg  ...   2001-01-01     0.0     0.0   
44410  /olEOCEGTW15WBPC0UnmLjokrmNF.jpg  ...   2016-03-12     0.0    76.0   
44576  /pErCRQoc3nt8lCsExd3fFYgyBox.jpg  ...   2016-01-27     0.0    84.0   
44655  /4VmQBE9qBRxCE3gHtSAa5AUsURq.jpg  ...   1897-01-01     0.0     1.0   

                                        spoken_languages    status tagline  \
19574  [{'iso_639_1': 'de', 'name': 'Deutsch'}, {'iso...  Released     NaN   
21602           [{'iso_639_1': 'en', 'name': 'English'}]  Released     NaN   
22832           [{'iso_639_1': 'en', 'name': 'English'}]  Released     NaN   
32141           [{'iso_639_1': 'en', 'name': 'English'}]  Released     NaN   
37407             [{'iso_639_1': 'cs', 'name': 'Český'}]  Released     NaN   
41047  [{'iso_639_1': 'ur', 'name': 'اردو'}, {'iso_63...  Released     NaN   
4

## 9.  original_title

In [42]:
data["original_title"].describe()

count      45424
unique     43363
top       Hamlet
freq           8
Name: original_title, dtype: object

In [43]:
pd.value_counts(data["original_title"])

Hamlet                         8
Alice in Wonderland            8
Macbeth                        7
A Christmas Carol              7
Cinderella                     7
                              ..
The Net 2.0                    1
The Little Kidnappers          1
Dude Bro Party Massacre III    1
Real Life                      1
The Evening Star               1
Name: original_title, Length: 43363, dtype: int64

In [44]:
data[data["original_title"]=="Alice in Wonderland"]

adult           belongs_to_collection     budget  \
1003   False                             NaN    3000000   
9665   False                             NaN   21000000   
11887  False                             NaN   14000000   
14892  False  Alice in Wonderland Collection  200000000   
15868  False                             NaN          0   
26894  False                             NaN          0   
32901  False                             NaN          0   
39530  False                             NaN          0   

                                                  genres     id  \
1003             [Animation, Adventure, Family, Fantasy]  12092   
9665                                   [Fantasy, Family]  30923   
11887                                  [Family, Fantasy]  34573   
14892                       [Family, Fantasy, Adventure]  12155   
15868  [Action, Adventure, Family, Fantasy, Science F...  25694   
26894                 [Family, Fantasy, Science Fiction]  35109   
32901                                  [Family, Fantasy]  87300   
39530                                          [Fantasy]  41278   

      original_language       original_title  \
1003                 en  Alice in Wonderland   
9665                 en  Alice in Wonderland   
11887                en  Alice in Wonderland   
14892                en  Alice in Wonderland   
15868                en  Alice in Wonderland   
26894                en  Alice in Wonderland   
32901                en  Alice in Wonderland   
39530                en  Alice in Wonderland   

                                                overview popularity  \
1003   On a golden afternoon, young Alice follows a W...    16.4889   
9665   Alice follows a white rabbit down a rabbit-hol...     8.6143   
11887  Classic tale of a girl named Alice who follows...    3.79413   
14892  Alice, an unpretentious and individual 19-year...    17.2851   
15868  On a boring winter afternoon, Alice dreams, th...    1.50435   
26894  Alice in Wonderland (1966) is a BBC television...    1.71604   
32901  A German adaptation of the classic Lewis Carro...   0.140291   
39530  This is the first movie version of the famous ...   0.708323   

                            poster_path  ... release_date       revenue  \
1003   /9gJa2B6M8jt85aDkb0MF98MLZ9T.jpg  ...   1951-07-03  5.720000e+08   
9665   /c53p3QRI0BNx7KE1zH8fXoWq4sM.jpg  ...   1999-02-28  0.000000e+00   
11887  /1jlw8lSaLetDP4XqioTgK6R38Lt.jpg  ...   1985-12-09  0.000000e+00   
14892  /pvEE5EN5N1yjmHmldfL4aJWm56l.jpg  ...   2010-03-03  1.025491e+09   
15868  /hA536GMozYb9zbMjo2NqESniveU.jpg  ...   1933-12-22  0.000000e+00   
26894  /dWnbHEZ31kMOx8cnN7IvG3fBpwp.jpg  ...   1966-12-26  0.000000e+00   
32901  /9YJqY2MWOUzy60iMhniOF4NEZRs.jpg  ...   1915-01-15  0.000000e+00   
39530  /rATc0e78ORRmJyvgqAywkejdgqh.jpg  ...   1903-10-17  0.000000e+00   

      runtime                              spoken_languages    status  \
1003     75.0      [{'iso_639_1': 'en', 'name': 'English'}]  Released   
9665    150.0      [{'iso_639_1': 'en', 'name': 'English'}]  Released   
11887   187.0      [{'iso_639_1': 'en', 'name': 'English'}]  Released   
14892   108.0      [{'iso_639_1': 'en', 'name': 'English'}]  Released   
15868    76.0      [{'iso_639_1': 'en', 'name': 'English'}]  Released   
26894    72.0      [{'iso_639_1': 'en', 'name': 'English'}]  Released   
32901    52.0      [{'iso_639_1': 'en', 'name': 'English'}]  Released   
39530     9.0  [{'iso_639_1': 'xx', 'name': 'No Language'}]  Released   

                                        tagline                title  video  \
1003    A world of wonders in One Great Picture  Alice in Wonderland  False   
9665            A Masterpiece of Imagination...  Alice in Wonderland  False   
11887                                       NaN  Alice in Wonderland  False   
14892  You're invited to a very important date.  Alice in Wonderland  False   
15868   The Entertainment Miracle Of All Times!  Alice in Wonderland 

## 10. overview

In [45]:
data["overview"].describe()

count                  44470
unique                 44297
top       No overview found.
freq                     133
Name: overview, dtype: object

In [46]:
data["overview"].value_counts()

No overview found.                                                                                                                                                                                                                                                                                                                                                                                                                                   133
No Overview                                                                                                                                                                                                                                                                                                                                                                                                                                            7
                                                                                                                      

In [47]:
text_counts = data["overview"].value_counts()
text_list = text_counts[text_counts > 1].index.tolist()

In [48]:
overview_list=[]
for overview in data["overview"]:
    if overview=="No overview found.":
        overview_list.append("")
    elif overview=="No Overview":
        overview_list.append("")
    elif overview==" ":
        overview_list.append("")
    elif overview=="No movie overview available.":
        overview_list.append("")
    else:
        overview_list.append(overview)

In [49]:
data["overview"]=overview_list

## 11. popularity
type을 object에서 numeric으로 변환

In [50]:
data["popularity"].describe()

count     45421
unique    44149
top         0.0
freq         34
Name: popularity, dtype: object

In [51]:
data["popularity"]=pd.to_numeric(data["popularity"],errors='coerce')

In [52]:
data["popularity"].describe()

count    45421.000000
mean         2.921493
std          6.007204
min          0.000000
25%          0.385850
50%          1.127098
75%          3.678394
max        547.488298
Name: popularity, dtype: float64

In [53]:
del data["poster_path"]

13. production_companies

In [54]:
data["production_companies"].describe()

count     45421
unique    22704
top          []
freq      11865
Name: production_companies, dtype: object

In [55]:

companies=[]
for string in data["production_companies"]:
    if type(string)==float:
        companies.append(float("Nan"))
    else:
        companies_list=eval(string)
        if len(companies_list)==0:
            companies.append(float("Nan"))
        else:
            companies2=[]
            for companies_dict in companies_list:
                companies_name=companies_dict["name"]
                companies2.append(companies_name)
            companies.append(companies2)

In [56]:
data["production_companies"]=companies

## 14. production_countries

In [57]:
data["production_countries"].describe()

count                                                 45421
unique                                                 2390
top       [{'iso_3166_1': 'US', 'name': 'United States o...
freq                                                  17838
Name: production_countries, dtype: object

In [58]:
data["production_countries"]

0        [{'iso_3166_1': 'US', 'name': 'United States o...
1        [{'iso_3166_1': 'US', 'name': 'United States o...
2        [{'iso_3166_1': 'US', 'name': 'United States o...
3        [{'iso_3166_1': 'US', 'name': 'United States o...
4        [{'iso_3166_1': 'US', 'name': 'United States o...
                               ...                        
45461               [{'iso_3166_1': 'IR', 'name': 'Iran'}]
45462        [{'iso_3166_1': 'PH', 'name': 'Philippines'}]
45463    [{'iso_3166_1': 'US', 'name': 'United States o...
45464             [{'iso_3166_1': 'RU', 'name': 'Russia'}]
45465     [{'iso_3166_1': 'GB', 'name': 'United Kingdom'}]
Name: production_countries, Length: 45424, dtype: object

In [59]:
countries=[]
for string in data["production_countries"]:
    if type(string)==float:
        countries.append(float("Nan"))
    else:
        countries_list=eval(string)
        if len(countries_list)==0:
            countries.append(float("Nan"))
        else:
            countries2=[]
            for countries_dict in countries_list:
                countries_name=countries_dict["name"]
                countries2.append(countries_name)
            countries.append(countries2)

In [60]:
data["production_countries"]=countries

In [61]:
data["production_countries"]

0        [United States of America]
1        [United States of America]
2        [United States of America]
3        [United States of America]
4        [United States of America]
                    ...            
45461                        [Iran]
45462                 [Philippines]
45463    [United States of America]
45464                      [Russia]
45465              [United Kingdom]
Name: production_countries, Length: 45424, dtype: object

## 15. release_date
datetime으로 변환

In [62]:
data["release_date"].describe()

count          45338
unique         17329
top       2008-01-01
freq             136
Name: release_date, dtype: object

In [63]:
from datetime import datetime

In [64]:
date=[]
for string in data["release_date"]:
    if type(string)!=str:
        date.append(float("Nan"))
    else:
        date.append(datetime.strptime(string, "%Y-%m-%d"))

In [65]:
data["release_date"]=date

In [66]:
data["release_date"]

0       1995-10-30
1       1995-12-15
2       1995-12-22
3       1995-12-22
4       1995-02-10
           ...    
45461          NaT
45462   2011-11-17
45463   2003-08-01
45464   1917-10-21
45465   2017-06-09
Name: release_date, Length: 45424, dtype: datetime64[ns]

## 16. revenue

In [67]:
data["revenue"].describe()

count    4.542100e+04
mean     1.121511e+07
std      6.435831e+07
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      2.787965e+09
Name: revenue, dtype: float64

In [68]:
np.sum(data["revenue"].isnull())

3

In [69]:
data[pd.isnull(data["revenue"])==True]
#얘네는 Nan이 여러개라서 없애는게 나을 것 같다고 함

adult        belongs_to_collection  budget  \
19729  False                          NaN       0   
29502  False  Mardock Scramble Collection       0   
35586  False                          NaN       0   

                                            genres      id original_language  \
19729                    [Action, Thriller, Drama]   82663                en   
29502                 [Animation, Science Fiction]  122662                ja   
35586  [TV Movie, Action, Horror, Science Fiction]  249260                en   

         original_title                                           overview  \
19729      Midnight Man  British soldiers force a recently captured IRA...   
29502  マルドゥック・スクランブル 排気         Third film of the Mardock Scramble series.   
35586  Avalanche Sharks  A group of skiers are terrorized during spring...   

       popularity production_companies  ... release_date revenue  runtime  \
19729         NaN                  NaN  ...          NaT     NaN      NaN   
29502         NaN                  NaN  ...          NaT     NaN      NaN   
35586         NaN                  NaN  ...          NaT     NaN      NaN   

       spoken_languages status tagline title video vote_average  vote_count  
19729               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
29502               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
35586               NaN    NaN     NaN   NaN   NaN          NaN         NaN  

[3 rows x 21 columns]

## 17. runtime

In [70]:
data["runtime"].describe()

count    45164.000000
mean        94.129196
std         38.413954
min          0.000000
25%         85.000000
50%         95.000000
75%        107.000000
max       1256.000000
Name: runtime, dtype: float64

In [71]:
np.sum(pd.isnull(data["runtime"]))

260

In [72]:
data[pd.isnull(data["runtime"])==True]
# Nan값이 많아서 그냥 둠. 0으로 대체하면 나중에 분석할 때 값이 달라질 것 같다고 함 

adult belongs_to_collection  budget          genres      id  \
634    False                   NaN       0        [Comedy]  287305   
635    False                   NaN       0        [Comedy]  339428   
644    False                   NaN       0         [Drama]  278978   
802    False                   NaN       0         [Drama]  282919   
863    False                   NaN       0      [Thriller]  253632   
...      ...                   ...     ...             ...     ...   
45246  False                   NaN       0        [Comedy]  231216   
45310  False                   NaN       0             NaN  418757   
45313  False                   NaN       0         [Drama]  369444   
45377  False                   NaN       0     [Adventure]  317389   
45410  False                   NaN       0  [Drama, Crime]  449131   

      original_language                  original_title overview  popularity  \
634                  de  Peanuts – Die Bank zahlt alles      NaN    0.066123   
635                  de                   Happy Weekend      NaN    0.002229   
644                  de       Und keiner weint mir nach      NaN    0.439989   
802                  de                       Diebinnen      NaN    0.106345   
863                  fr                     Baton Rouge      NaN    0.437895   
...                 ...                             ...      ...         ...   
45246                de                 Villa Henriette      NaN    0.002513   
45310                pl         Między nami dobrze jest      NaN    0.030803   
45313                pl                 Ostatni dzwonek      NaN    0.000102   
45377                es   Simbad e il califfo di Bagdad      NaN    0.006352   
45410                ru                           Aprel      NaN    0.008903   

                 production_companies  ... release_date  revenue  runtime  \
634    [Westdeutscher Rundfunk (WDR)]  ...   1996-03-21      0.0      NaN   
635         [Senator Film Produktion]  ...   1996-03-14  65335.0      NaN   
644                               NaN  ...   1996-02-29      0.0      NaN   
802                               NaN  ...   1996-06-20      0.0      NaN   
863                               NaN  ...   1988-10-08      0.0      NaN   
...                               ...  ...          ...      ...      ...   
45246                             NaN  ...   2004-12-02      0.0      NaN   
45310                             NaN  ...   2014-08-01      0.0      NaN   
45313                             NaN  ...   1989-10-27      0.0      NaN   
45377   [Roas Produzioni, Buton Film]  ...   1973-07-22      0.0      NaN   
45410                             NaN  ...          NaT      0.0      NaN   

                                spoken_languages    status tagline  \
634                                           []  Released     NaN   
635     [{'iso_639_1': 'de', 'name': 'Deutsch'}]  Released     NaN   
644     [{'iso_639_1': 'de', 'name': 'Deutsch'}]  Released     NaN   
802                                           []  Released     NaN   
863                                           []  Released     NaN   
...                                          ...       ...     ...   
45246   [{'iso_639_1': 'de', 'name': 'Deutsch'}]  Released     NaN   
45310                                         []  Released     NaN   
45313                                         []  Released     NaN   
45377  [{'iso_639_1': 'it', 'name': 'Italiano'}]  Released     NaN   
45410                                         []  Released     NaN   

                                title  video vote_average  vote_count  
634    Peanuts – Die Bank zahlt alles  False          4.0         1.0  
635                     Happy Weekend  False          0.0         0.0  
644         Und keiner weint mir nach  False          0.0         0.0  
802                         Diebinnen  False          4.0         1.0  
863                       Baton Rouge  False          0.0         0.0  
...    

## 18. spoken_languages

In [73]:
data["spoken_languages"].describe()

count                                        45421
unique                                        1931
top       [{'iso_639_1': 'en', 'name': 'English'}]
freq                                         22375
Name: spoken_languages, dtype: object

In [74]:
np.sum(pd.isnull(data["spoken_languages"]))

3

In [75]:
data[pd.isnull(data["spoken_languages"])==True]

adult        belongs_to_collection  budget  \
19729  False                          NaN       0   
29502  False  Mardock Scramble Collection       0   
35586  False                          NaN       0   

                                            genres      id original_language  \
19729                    [Action, Thriller, Drama]   82663                en   
29502                 [Animation, Science Fiction]  122662                ja   
35586  [TV Movie, Action, Horror, Science Fiction]  249260                en   

         original_title                                           overview  \
19729      Midnight Man  British soldiers force a recently captured IRA...   
29502  マルドゥック・スクランブル 排気         Third film of the Mardock Scramble series.   
35586  Avalanche Sharks  A group of skiers are terrorized during spring...   

       popularity production_companies  ... release_date revenue  runtime  \
19729         NaN                  NaN  ...          NaT     NaN      NaN   
29502         NaN                  NaN  ...          NaT     NaN      NaN   
35586         NaN                  NaN  ...          NaT     NaN      NaN   

       spoken_languages status tagline title video vote_average  vote_count  
19729               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
29502               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
35586               NaN    NaN     NaN   NaN   NaN          NaN         NaN  

[3 rows x 21 columns]

In [76]:
language=[]
for string in data["spoken_languages"]:
    if type(string)!=str:
        language.append(float("Nan"))
    else:
        language_list=eval(string)
        if len(language_list)==0:
            language.append(float("Nan"))
        else:
            language2=[]
            for language_dict in language_list:
                language_name=language_dict["name"]
                language2.append(language_name)
            language.append(language2)

In [77]:
data["spoken_languages"]=language

In [78]:
data["spoken_languages"]

0                  [English]
1        [English, Français]
2                  [English]
3                  [English]
4                  [English]
                ...         
45461                [فارسی]
45462                     []
45463              [English]
45464                    NaN
45465              [English]
Name: spoken_languages, Length: 45424, dtype: object

In [79]:
for string in data["spoken_languages"]:
    if type(string)==list:
        for lan in string:
            if lan=="":
                string.remove("")

In [80]:
data["spoken_languages"]

0                  [English]
1        [English, Français]
2                  [English]
3                  [English]
4                  [English]
                ...         
45461                [فارسی]
45462                     []
45463              [English]
45464                    NaN
45465              [English]
Name: spoken_languages, Length: 45424, dtype: object

## 19. status

In [81]:
data["status"].describe()

count        45340
unique           6
top       Released
freq         44977
Name: status, dtype: object

In [82]:
pd.value_counts(data["status"])

Released           44977
Rumored              229
Post Production       98
In Production         19
Planned               15
Canceled               2
Name: status, dtype: int64

In [83]:
np.sum(pd.isnull(data["status"]))

84

In [84]:
# Nan인 애들 중에 popularity나 revenue가 0이 아닌 것들은 개봉한 것으로 생각할 수 있으니, released로 채운다고 함
data[pd.isnull(data["status"])==True]

adult belongs_to_collection  budget                     genres      id  \
189    False                   NaN       0              [Documentary]   56088   
767    False                   NaN       0              [Documentary]   52059   
1384   False                   NaN       0                   [Comedy]   40506   
2140   False                   NaN       0                        NaN   77314   
3460   False                   NaN       0                    [Drama]   94214   
...      ...                   ...     ...                        ...     ...   
44377  False                   NaN       0                        NaN   39827   
45067  False                   NaN       0  [Drama, Romance, Foreign]   49492   
45090  False                   NaN       0                        NaN  128268   
45159  False                   NaN       0                        NaN   41811   
45442  False                   NaN       0                        NaN   44330   

      original_language                                 original_title  \
189                  en                                       The Show   
767                  en                     The Gate of Heavenly Peace   
1384                 en                              Meet Wally Sparks   
2140                 fr                     The Cabinet of Dr. Ramirez   
3460                 en                     Jails, Hospitals & Hip-Hop   
...                 ...                                            ...   
44377                en           The Who: Thirty Years of Maximum R&B   
45067                en                          Jogwa - The Awakening   
45090                en  Veruschka - Die Inszenierung (m)eines Körpers   
45159                en                                         Eating   
45442                en                           Le tripot clandestin   

                                                overview  popularity  \
189    A documentary about the culture of hip-hop. Th...    0.159596   
767    The Gate of Heavenly Peace is a feature-length...    0.107050   
1384   Wally Sparks is a tabloid TV show reporter who...    0.576741   
2140                                                 NaN    0.037500   
3460   Jails, Hospitals &amp; Hip-Hop is a cinematic ...    0.009057   
...                                                  ...         ...   
44377              A celebration of 30 years of The Who.    0.014260   
45067  Jogwa actually means alms given to a person, u...    0.004990   
45090         Documentary about Veruschka von Lehndorff.    0.056298   
45159  At a spacious house in Los Angeles, Helene is ...    0.000002   
45442  A combination gambling den and bawdy house is ...    0.219260   

      production_companies  ... release_date revenue  runtime  \
189                    NaN  ...   1995-08-25     0.0     93.0   
767                    NaN  ...   1995-01-01     0.0    180.0   
1384                   NaN  ...   1997-01-01     0.0    105.0   
2140                   NaN  ...   1991-12-04     0.0      NaN   
3460                   NaN  ...          NaT    10.0     90.0   
...                    ...  ...          ...     ...      ...   
44377                  NaN  ...   1994-08-15     0.0    155.0   
45067                  NaN  ...   2009-09-25     0.0     92.0   
45090                  NaN  ...   2005-09-30     0.0     52.0   
45159                  NaN  ...   1990-11-30     0.0    110.0   
45442                  NaN  ...   1905-01-01     0.0      3.0   

          spoken_languages status  \
189                    NaN    NaN   
767                    NaN    NaN   
1384             [English]    NaN   
2140                   NaN    NaN   
3460                   NaN    NaN   
...                    ...    ...   
44377                  NaN    NaN   
45067                  NaN    NaN   
45090  [Deutsch, Français]    NaN   
45159            [English]    NaN   
45442                  NaN    NaN   

                                                 tagli

## 20. tagline

In [85]:
data["tagline"].describe()

count                      20395
unique                     20277
top       Based on a true story.
freq                           7
Name: tagline, dtype: object

In [86]:
pd.value_counts(data["tagline"])

Based on a true story.                                                                                                                7
-                                                                                                                                     4
Be careful what you wish for.                                                                                                         4
Trust no one.                                                                                                                         4
Documentary                                                                                                                           3
                                                                                                                                     ..
The world will never be the same, once you've seen it through the eyes of Forrest Gump.                                               1
The story of hot-blooded Maria... who always get

In [87]:
np.sum(pd.isnull(data["tagline"]))

25029

In [88]:
data[pd.isnull(data["tagline"])==True]

adult belongs_to_collection    budget                       genres  \
0      False  Toy Story Collection  30000000  [Animation, Comedy, Family]   
11     False                   NaN         0             [Comedy, Horror]   
27     False                   NaN         0             [Drama, Romance]   
32     False                   NaN         0         [Romance, Adventure]   
35     False                   NaN  11000000                      [Drama]   
...      ...                   ...       ...                          ...   
45459  False                   NaN         0            [Science Fiction]   
45460  False                   NaN         0     [Drama, Action, Romance]   
45462  False                   NaN         0                      [Drama]   
45464  False                   NaN         0                          NaN   
45465  False                   NaN         0                          NaN   

           id original_language                    original_title  \
0         862                en                         Toy Story   
11      12110                en       Dracula: Dead and Loving It   
27      17015                en                        Persuasion   
32      78802                fr  Guillaumet, les ailes du courage   
35        687                en                  Dead Man Walking   
...       ...               ...                               ...   
45459  222848                en                   Caged Heat 3000   
45460   30840                en                        Robin Hood   
45462  111109                tl               Siglo ng Pagluluwal   
45464  227506                en               Satana likuyushchiy   
45465  461257                en                          Queerama   

                                                overview  popularity  \
0      Led by Woody, Andy's toys live happily in his ...   21.946943   
11     When a lawyer shows up at the vampire's doorst...    5.430331   
27     This film adaptation of Jane Austen's last nov...    2.228434   
32                                                   NaN    0.745542   
35     A justice drama based on a true story about a ...    6.891317   
...                                                  ...         ...   
45459  It's the year 3000 AD. The world's most danger...    0.661558   
45460  Yet another version of the classic epic, with ...    5.683753   
45462  An artist struggles to finish his work while a...    0.178241   
45464  In a small town live two brothers, one a minis...    0.003503   
45465  50 years after decriminalisation of homosexual...    0.163015   

                                    production_companies  ... release_date  \
0                              [Pixar Animation Studios]  ...   1995-10-30   
11     [Columbia Pictures, Castle Rock Entertainment,...  ...   1995-12-22   
27                                           [BBC Films]  ...   1995-09-27   
32                                [Iwerks Entertainment]  ...   1996-09-18   
35     [Havoc, PolyGram Filmed Entertainment, Working...  ...   1995-12-29   
...                                                  ...  ...          ...   
45459                            [Concorde-New Horizons]  ...   1995-01-01   
45460  [Westdeutscher Rundfunk (WDR), Working Title F...  ...   1991-05-13   
45462                                      [Sine Olivia]  ...   2011-11-17   
45464                                        [Yermoliev]  ...   1917-10-21   
45465                                                NaN  ...   2017-06-09   

           revenue  runtime    spoken_languages    status tagline  \
0      373554033.0     81.0           [English]  Released     NaN   
11             0.0     88.0  [English, Deutsch]  Released     NaN   
27             0.0    104.0           [English]  Released     NaN   
32             0.0     50.0           [English]  Released     NaN   
35      39363635.0    122.0           [English]  Released     NaN   
...            ...      ...          

## 21. title

In [89]:
data["title"].describe()

count          45421
unique         42269
top       Cinderella
freq              11
Name: title, dtype: object

In [90]:
pd.value_counts(data["title"])

Cinderella                 11
Hamlet                      9
Alice in Wonderland         9
Beauty and the Beast        8
Les Misérables              8
                           ..
India Song                  1
The Youngest Profession     1
Surf's Up                   1
Archipelago                 1
The Evening Star            1
Name: title, Length: 42269, dtype: int64

In [91]:
np.sum(pd.isnull(data["title"]))

3

In [92]:
data[pd.isnull(data["title"])==True]

adult        belongs_to_collection  budget  \
19729  False                          NaN       0   
29502  False  Mardock Scramble Collection       0   
35586  False                          NaN       0   

                                            genres      id original_language  \
19729                    [Action, Thriller, Drama]   82663                en   
29502                 [Animation, Science Fiction]  122662                ja   
35586  [TV Movie, Action, Horror, Science Fiction]  249260                en   

         original_title                                           overview  \
19729      Midnight Man  British soldiers force a recently captured IRA...   
29502  マルドゥック・スクランブル 排気         Third film of the Mardock Scramble series.   
35586  Avalanche Sharks  A group of skiers are terrorized during spring...   

       popularity production_companies  ... release_date revenue  runtime  \
19729         NaN                  NaN  ...          NaT     NaN      NaN   
29502         NaN                  NaN  ...          NaT     NaN      NaN   
35586         NaN                  NaN  ...          NaT     NaN      NaN   

       spoken_languages status tagline title video vote_average  vote_count  
19729               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
29502               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
35586               NaN    NaN     NaN   NaN   NaN          NaN         NaN  

[3 rows x 21 columns]

## 22. video 
>> 변수가 뭘 의미하는지 모르겠어서 그냥 버리는게 나을 것 같긴 한데 우선 그냥 냅둠

In [93]:
data["video"].describe()

count     45421
unique        2
top       False
freq      45328
Name: video, dtype: object

In [94]:
pd.value_counts(data["video"])

False    45328
True        93
Name: video, dtype: int64

In [95]:
np.sum(pd.isnull(data["video"]))

3

In [96]:
data[pd.isnull(data["video"])==True]

adult        belongs_to_collection  budget  \
19729  False                          NaN       0   
29502  False  Mardock Scramble Collection       0   
35586  False                          NaN       0   

                                            genres      id original_language  \
19729                    [Action, Thriller, Drama]   82663                en   
29502                 [Animation, Science Fiction]  122662                ja   
35586  [TV Movie, Action, Horror, Science Fiction]  249260                en   

         original_title                                           overview  \
19729      Midnight Man  British soldiers force a recently captured IRA...   
29502  マルドゥック・スクランブル 排気         Third film of the Mardock Scramble series.   
35586  Avalanche Sharks  A group of skiers are terrorized during spring...   

       popularity production_companies  ... release_date revenue  runtime  \
19729         NaN                  NaN  ...          NaT     NaN      NaN   
29502         NaN                  NaN  ...          NaT     NaN      NaN   
35586         NaN                  NaN  ...          NaT     NaN      NaN   

       spoken_languages status tagline title video vote_average  vote_count  
19729               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
29502               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
35586               NaN    NaN     NaN   NaN   NaN          NaN         NaN  

[3 rows x 21 columns]

In [97]:
data[data["video"]==True]

adult    belongs_to_collection  budget  \
5492   False                      NaN       0   
11469  False  Loose Change Collection    5000   
13564  False                      NaN       0   
14271  False     Love Hina Collection       0   
14608  False                      NaN       0   
...      ...                      ...     ...   
44286  False                      NaN       0   
44798  False                      NaN       0   
44861  False                      NaN       0   
44981  False                      NaN       0   
45199  False                      NaN       0   

                                                  genres      id  \
5492   [Action, Adventure, Animation, Comedy, Family,...   15173   
11469                                      [Documentary]   14277   
13564                                      [Documentary]  284117   
14271                [Comedy, Romance, Animation, Music]   44685   
14608               [Action, Adventure, Family, Fantasy]   35148   
...                                                  ...     ...   
44286                        [Action, Animation, Horror]  161337   
44798               [Animation, Horror, Science Fiction]  425916   
44861                                                NaN  334795   
44981                            [Family, Drama, Comedy]  461088   
45199                                [Comedy, Animation]  279687   

      original_language                          original_title  \
5492                 en              Jonah: A VeggieTales Movie   
11469                en               Loose Change: 2nd Edition   
13564                en                                   Fitna   
14271                ja              ラブひな 春スペシャル ～キミサクラチルナカレ!!～   
14608                en  Playmobil: The Secret of Pirate Island   
...                 ...                                     ...   
44286                en                 Underworld: Endless War   
44798                en                           Salad Fingers   
44861                ru                           Конец Каникул   
44981                fa                          ۵۰ کیلو آلبالو   
45199                en            Futurama: The Lost Adventure   

                                                overview  popularity  \
5492   Get ready as Bob the Tomato, Larry the Cucumbe...    1.734181   
11469  What if...September 11th was not a surprise at...    0.651714   
13564  A short film in which Quran verses are shown a...    0.040068   
14271  The tests have come, and Keitaro can only hope...    1.785708   
14608  Ahoy, matey! Your Favorite Playmobil toys have...    0.014168   
...                                                  ...         ...   
44286  Underworld: Endless War is a collection of thr...    3.301063   
44798  The adventures of Salad Fingers, an unspecifie...    0.141367   
44861  Four music videos of Kino band joined together...    0.000000   
44981  When the corrupted groom starts shooting in th...    0.034119   
45199  The cut-scenes from Futurama (2003) (VG), edit...    0.988667   

             production_companies  ... release_date revenue  runtime  \
5492       [Big Idea Productions]  ...   2002-10-04     0.0     82.0   
11469  [Polar Film   Medien GmbH]  ...   2006-06-01     0.0    100.0   
13564                         NaN  ...   2008-03-27     0.0     15.0   
14271           [TV Tokyo, Xebec]  ...   2001-03-31     0.0     48.0   
14608                         NaN  ...   2009-10-13     0.0    140.0   
...                           ...  ...          ...     ...      ...   
44286                         NaN  ...   2011-10-19     0.0     18.0   
44798                         NaN  ...          NaT     0.0     52.0   
44861                         NaN  ...   1986-01-01     0.0     23.0   
44981                         NaN  ...   2016-03-22     0.0     89.0   
45199     [The Curiosity Company]  ...   2008-06-24     0.0     30.0   

                                        spoken_languages    status

## 23. vote_average

In [98]:
data["vote_average"].describe()

count    45421.000000
mean         5.618665
std          1.924008
min          0.000000
25%          5.000000
50%          6.000000
75%          6.800000
max         10.000000
Name: vote_average, dtype: float64

In [99]:
np.sum(pd.isnull(data["vote_average"]))

3

In [100]:
data[pd.isnull(data["vote_average"])==True]

adult        belongs_to_collection  budget  \
19729  False                          NaN       0   
29502  False  Mardock Scramble Collection       0   
35586  False                          NaN       0   

                                            genres      id original_language  \
19729                    [Action, Thriller, Drama]   82663                en   
29502                 [Animation, Science Fiction]  122662                ja   
35586  [TV Movie, Action, Horror, Science Fiction]  249260                en   

         original_title                                           overview  \
19729      Midnight Man  British soldiers force a recently captured IRA...   
29502  マルドゥック・スクランブル 排気         Third film of the Mardock Scramble series.   
35586  Avalanche Sharks  A group of skiers are terrorized during spring...   

       popularity production_companies  ... release_date revenue  runtime  \
19729         NaN                  NaN  ...          NaT     NaN      NaN   
29502         NaN                  NaN  ...          NaT     NaN      NaN   
35586         NaN                  NaN  ...          NaT     NaN      NaN   

       spoken_languages status tagline title video vote_average  vote_count  
19729               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
29502               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
35586               NaN    NaN     NaN   NaN   NaN          NaN         NaN  

[3 rows x 21 columns]

## 24. vote_count

In [101]:
data["vote_count"].describe()

count    45421.000000
mean       109.956210
std        491.512921
min          0.000000
25%          3.000000
50%         10.000000
75%         34.000000
max      14075.000000
Name: vote_count, dtype: float64

In [102]:
np.sum(pd.isnull(data["vote_count"]))

3

In [103]:
data[pd.isnull(data["vote_count"])==True]

adult        belongs_to_collection  budget  \
19729  False                          NaN       0   
29502  False  Mardock Scramble Collection       0   
35586  False                          NaN       0   

                                            genres      id original_language  \
19729                    [Action, Thriller, Drama]   82663                en   
29502                 [Animation, Science Fiction]  122662                ja   
35586  [TV Movie, Action, Horror, Science Fiction]  249260                en   

         original_title                                           overview  \
19729      Midnight Man  British soldiers force a recently captured IRA...   
29502  マルドゥック・スクランブル 排気         Third film of the Mardock Scramble series.   
35586  Avalanche Sharks  A group of skiers are terrorized during spring...   

       popularity production_companies  ... release_date revenue  runtime  \
19729         NaN                  NaN  ...          NaT     NaN      NaN   
29502         NaN                  NaN  ...          NaT     NaN      NaN   
35586         NaN                  NaN  ...          NaT     NaN      NaN   

       spoken_languages status tagline title video vote_average  vote_count  
19729               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
29502               NaN    NaN     NaN   NaN   NaN          NaN         NaN  
35586               NaN    NaN     NaN   NaN   NaN          NaN         NaN  

[3 rows x 21 columns]

## Nan이 많은 세 행 제거

In [104]:
data=data.drop([19729, 29502, 35586])

In [105]:
data[pd.isnull(data["vote_count"])==True]

Empty DataFrame
Columns: [adult, belongs_to_collection, budget, genres, id, original_language, original_title, overview, popularity, production_companies, production_countries, release_date, revenue, runtime, spoken_languages, status, tagline, title, video, vote_average, vote_count]
Index: []

[0 rows x 21 columns]

### original_language 다시 정리

In [106]:
data["original_language"].describe()

count     45410
unique       89
top          en
freq      32244
Name: original_language, dtype: object

In [107]:
np.sum(pd.isnull(data["original_language"]))

11

In [108]:
index = pd.isnull(data["original_language"])==True

In [109]:
data.loc[index,["original_language","original_title","production_countries","spoken_languages","title","release_date"]]

original_language              original_title  \
19574               NaN     Shadowing the Third Man   
21602               NaN              Unfinished Sky   
22832               NaN             13 Fighting Men   
32141               NaN                   Lambchops   
37407               NaN                      Bajaja   
41047               NaN              Song of Lahore   
41872               NaN  Annabelle Serpentine Dance   
44057               NaN       Lettre d'une inconnue   
44410               NaN                        Garn   
44576               NaN                      WiNWiN   
44655               NaN       La prise de Tournavos   

                                    production_countries  \
19574  [Austria, France, Japan, United Kingdom, Unite...   
21602                                        [Australia]   
22832                         [United States of America]   
32141                         [United States of America]   
37407                                   [Czech Republic]   
41047                                                NaN   
41872                         [United States of America]   
44057                                           [France]   
44410                                                NaN   
44576                                          [Austria]   
44655                                           [France]   

                   spoken_languages                       title release_date  
19574            [Deutsch, English]     Shadowing the Third Man   2004-10-11  
21602                     [English]              Unfinished Sky   2007-08-04  
22832                     [English]             13 Fighting Men   1960-03-31  
32141                     [English]                   Lambchops   1929-10-31  
37407                       [Český]               Prince Bayaya   1950-01-26  
41047       [اردو, English, ਪੰਜਾਬੀ]              Song of Lahore   2015-11-13  
41872                 [No Language]  Annabelle Serpentine Dance   1895-04-01  
44057                    [Français]       Lettre d'une inconnue   2001-01-01  
44410  [svenska, English, Íslenska]                        Yarn   2016-03-12  
44576            [Deutsch, English]                      WiNWiN   2016-01-27  
44655                 [No Language]  The Surrender of Tournavos   1897-01-01

In [110]:
data.loc[[19574,21602,22832,32141,41047,41872,44576,44655],"original_language"]="en"

In [111]:
data.loc[37407,"original_language"]="cs"

In [112]:
data.loc[44057,"original_language"]="fr"

In [113]:
data.loc[44410,"original_language"]="sv"

In [114]:
data.loc[index,["original_language","original_title","production_countries","spoken_languages","title","release_date"]]

original_language              original_title  \
19574                en     Shadowing the Third Man   
21602                en              Unfinished Sky   
22832                en             13 Fighting Men   
32141                en                   Lambchops   
37407                cs                      Bajaja   
41047                en              Song of Lahore   
41872                en  Annabelle Serpentine Dance   
44057                fr       Lettre d'une inconnue   
44410                sv                        Garn   
44576                en                      WiNWiN   
44655                en       La prise de Tournavos   

                                    production_countries  \
19574  [Austria, France, Japan, United Kingdom, Unite...   
21602                                        [Australia]   
22832                         [United States of America]   
32141                         [United States of America]   
37407                                   [Czech Republic]   
41047                                                NaN   
41872                         [United States of America]   
44057                                           [France]   
44410                                                NaN   
44576                                          [Austria]   
44655                                           [France]   

                   spoken_languages                       title release_date  
19574            [Deutsch, English]     Shadowing the Third Man   2004-10-11  
21602                     [English]              Unfinished Sky   2007-08-04  
22832                     [English]             13 Fighting Men   1960-03-31  
32141                     [English]                   Lambchops   1929-10-31  
37407                       [Český]               Prince Bayaya   1950-01-26  
41047       [اردو, English, ਪੰਜਾਬੀ]              Song of Lahore   2015-11-13  
41872                 [No Language]  Annabelle Serpentine Dance   1895-04-01  
44057                    [Français]       Lettre d'une inconnue   2001-01-01  
44410  [svenska, English, Íslenska]                        Yarn   2016-03-12  
44576            [Deutsch, English]                      WiNWiN   2016-01-27  
44655                 [No Language]  The Surrender of Tournavos   1897-01-01

In [115]:
#############################
###Demographic recommender###
#############################
####Using Weighted Rating###

In [116]:
data.head(3)

adult      belongs_to_collection    budget                        genres  \
0  False       Toy Story Collection  30000000   [Animation, Comedy, Family]   
1  False                        NaN  65000000  [Adventure, Fantasy, Family]   
2  False  Grumpy Old Men Collection         0             [Romance, Comedy]   

      id original_language    original_title  \
0    862                en         Toy Story   
1   8844                en           Jumanji   
2  15602                en  Grumpier Old Men   

                                            overview  popularity  \
0  Led by Woody, Andy's toys live happily in his ...   21.946943   
1  When siblings Judy and Peter discover an encha...   17.015539   
2  A family wedding reignites the ancient feud be...   11.712900   

                                production_companies  ... release_date  \
0                          [Pixar Animation Studios]  ...   1995-10-30   
1  [TriStar Pictures, Teitler Film, Interscope Co...  ...   1995-12-15   
2                     [Warner Bros., Lancaster Gate]  ...   1995-12-22   

       revenue  runtime     spoken_languages    status  \
0  373554033.0     81.0            [English]  Released   
1  262797249.0    104.0  [English, Français]  Released   
2          0.0    101.0            [English]  Released   

                                             tagline             title  video  \
0                                                NaN         Toy Story  False   
1          Roll the dice and unleash the excitement!           Jumanji  False   
2  Still Yelling. Still Fighting. Still Ready for...  Grumpier Old Men  False   

  vote_average  vote_count  
0          7.7      5415.0  
1          6.9      2413.0  
2          6.5        92.0  

[3 rows x 21 columns]

In [117]:
m=data['vote_count'].quantile(q=0.9)

In [118]:
data_filtered = data[data['vote_count']>m]

In [119]:
data_filtered.shape

(4534, 21)

In [120]:
data['vote_count'].astype('int')

0        5415
1        2413
2          92
3          34
4         173
         ... 
45461       1
45462       3
45463       6
45464       0
45465       0
Name: vote_count, Length: 45421, dtype: int32

In [121]:
def movie_score(x):
    v=x['vote_count']
    m=data['vote_count'].quantile(q=0.9)
    R=x['vote_average']
    C=data['vote_average'].mean()
    return ((R*v)/(v+m))+((C*m)/(v+m))

In [122]:
data_filtered['score']=movie_score(data_filtered)

C:\Users\user\.conda\envs\test\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [123]:
highscore = data_filtered.sort_values(by='score', ascending=False).head(10)

In [124]:
highscore[['title', 'vote_count', 'vote_average', 'popularity', 'score']]

title  vote_count  vote_average  popularity  \
314       The Shawshank Redemption      8358.0           8.5   51.645403   
834                  The Godfather      6024.0           8.5   41.109264   
10309  Dilwale Dulhania Le Jayenge       661.0           9.1   34.457024   
12481              The Dark Knight     12269.0           8.3  123.167259   
2843                    Fight Club      9678.0           8.3   63.869599   
292                   Pulp Fiction      8670.0           8.3  140.950236   
522               Schindler's List      4436.0           8.3   41.725123   
23673                     Whiplash      4376.0           8.3   64.299990   
5481                 Spirited Away      3968.0           8.3   41.048867   
2211             Life Is Beautiful      3643.0           8.3   39.394970   

          score  
314    8.445878  
834    8.425451  
10309  8.421543  
12481  8.265483  
2843   8.256392  
292    8.251414  
522    8.206655  
23673  8.205420  
5481   8.196072  
2211   8.187191

In [125]:
data_popular = data.sort_values('popularity', ascending=False)

In [142]:
data_popular[['title', 'vote_count', 'vote_average', 'popularity']].head(10)

title  vote_count  vote_average  popularity
30700                         Minions      4729.0           6.4  547.488298
33356                    Wonder Woman      5025.0           7.2  294.337037
42222            Beauty and the Beast      5530.0           6.8  287.253654
43644                     Baby Driver      2083.0           7.2  228.032744
24455                      Big Hero 6      6289.0           7.8  213.849907
26564                        Deadpool     11444.0           7.4  187.860492
26566  Guardians of the Galaxy Vol. 2      4858.0           7.6  185.330992
14551                          Avatar     12114.0           7.2  185.070892
24351                       John Wick      5499.0           7.0  183.870374
23675                       Gone Girl      6023.0           7.9  154.801009

# (+) Credits.csv

In [127]:
cred = pd.read_csv('credits.csv')

In [128]:
cred["id"].describe()

count     45476.000000
mean     108345.997537
std      112443.796536
min           2.000000
25%       26443.250000
50%       60002.500000
75%      157302.000000
max      469172.000000
Name: id, dtype: float64

In [129]:
cred["id"]=pd.to_numeric(cred["id"],errors='coerce')

In [130]:
pd.value_counts(cred["id"])

141971    3
187156    2
152795    2
99080     2
24026     2
         ..
19164     1
29403     1
31450     1
25305     1
276895    1
Name: id, Length: 45432, dtype: int64

In [131]:
cred=cred.drop_duplicates(["id"],keep="first")

In [132]:
np.sum(cred.duplicated(["id"]))

0

In [133]:
mc = pd.merge(cred, data, on='id')

In [134]:
mc.head(2)

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   

                                                crew    id  adult  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   862  False   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...  8844  False   

  belongs_to_collection    budget                        genres  \
0  Toy Story Collection  30000000   [Animation, Comedy, Family]   
1                   NaN  65000000  [Adventure, Fantasy, Family]   

  original_language original_title  \
0                en      Toy Story   
1                en        Jumanji   

                                            overview  ...  release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...    1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...    1995-12-15   

       revenue runtime     spoken_languages    status  \
0  373554033.0    81.0            [English]  Released   
1  262797249.0   104.0  [English, Français]  Released   

                                     tagline      title  video vote_average  \
0                                        NaN  Toy Story  False          7.7   
1  Roll the dice and unleash the excitement!    Jumanji  False          6.9   

  vote_count  
0     5415.0  
1     2413.0  

[2 rows x 23 columns]

컨텐츠 기반 분석 예측

In [135]:
mc['overview'].head()

0    Led by Woody, Andy's toys live happily in his ...
1    When siblings Judy and Peter discover an encha...
2    A family wedding reignites the ancient feud be...
3    Cheated on, mistreated and stepped on, the wom...
4    Just when George Banks has recovered from his ...
Name: overview, dtype: object

NaN을 공백으로 바꾸기

In [136]:
mc['overview'].isnull().sum()

954

In [137]:
mc['overview'].fillna(' ', inplace=True)

TF-IDF Matrix 만들기

In [138]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import linear_kernel

In [139]:
# 램크기가 작아서 오류가 뜨는데 일단은 그냥 함
tfidfv = TfidfVectorizer(analyzer='word', stop_words='english')
tfidfv_matrix = tfidfv.fit_transform(mc['overview'])
tfidfv_matrix.todense().shape

(45420, 75824)

(45420, 75824) => 45420개의 영화를 설명하기 위해 75824개의 단어들이 사용되었다는 것을 알 수 있다.

Similarity Score 계산하기 -> 유사한 영화들을 추측할 수 있음

In [ ]:
cosine_sim1 = linear_kernel(tfidfv_matrix, tfidfv_matrix)
cosine_sim1.shape 
#이것도 램크기가 작아서 그런데, 어쨌든 (45420, 45420)의 정방형 코사인 similarity matrix가 완성됨

영화 제목 입력하면 비슷한 영화 10개 추측

In [144]:
indices=pd.Series(data=list(mc.index), index= mc['title'] )

In [145]:
indices.head()

title
Toy Story                      0
Jumanji                        1
Grumpier Old Men               2
Waiting to Exhale              3
Father of the Bride Part II    4
dtype: int64

In [146]:
# 영화 제목을 넣으면 비슷한 10개의 영화들을 예측(추천)하는 방식의 함수
def content_recommendations(title, cosine_sim):
    
    # 영화제목에 해당하는 인덱스를 설정
    idx = indices[title]
    
    # Pairwsie similarity scores 구하기
    sim_scores = list(enumerate(cosine_sim[idx]))
    
    # similarity scores로 영화 sort!
    sim_scores.sort(key=lambda x: x[1], reverse=True)

    # 비슷한 영화 10개 점수 구하기
    sim_scores=sim_scores[1:11]
    
    # 영화 인덱스 찾고
    ind=[]
    for (x,y) in sim_scores:
        ind.append(x)
        
    # Top 10 
    tit=[]
    for x in ind:
        tit.append(mc.iloc[x]['title'])
    return pd.Series(data=tit, index=ind)

In [124]:
mc['title']

0                          Toy Story
1                            Jumanji
2                   Grumpier Old Men
3                  Waiting to Exhale
4        Father of the Bride Part II
                    ...             
45415                         Subdue
45416            Century of Birthing
45417                       Betrayal
45418               Satan Triumphant
45419                       Queerama
Name: title, Length: 45420, dtype: object

In [150]:
content_recommendations("Father of the Bride Part II", cosine_sim1)

6792        Father of the Bride
6570                      Kuffs
6305            North to Alaska
19790                   Babbitt
34435         You're Killing Me
13606       The Magic of Méliès
5004                    Wendigo
27953          I Start Counting
43844    George of the Jungle 2
7096         The Out of Towners
dtype: object

In [149]:
content_recommendations('Grumpier Old Men',cosine_sim1)

9205     An Extremely Goofy Movie
35544                         Max
443                      Fearless
235                 A Goofy Movie
4100                Heartbreakers
24557               The Guardians
31678        The Phantom of Paris
1616                         Bent
2281                     Rushmore
35273            The Zohar Secret
dtype: object

♣ 심화적으로 영화 감독, 캐스팅, 장르에 기반한 분석을 추가한다.

Cast 전처리

In [154]:
mc['cast']

0        [{'cast_id': 14, 'character': 'Woody (voice)',...
1        [{'cast_id': 1, 'character': 'Alan Parrish', '...
2        [{'cast_id': 2, 'character': 'Max Goldman', 'c...
3        [{'cast_id': 1, 'character': "Savannah 'Vannah...
4        [{'cast_id': 1, 'character': 'George Banks', '...
                               ...                        
45415    [{'cast_id': 0, 'character': '', 'credit_id': ...
45416    [{'cast_id': 1002, 'character': 'Sister Angela...
45417    [{'cast_id': 6, 'character': 'Emily Shaw', 'cr...
45418    [{'cast_id': 2, 'character': '', 'credit_id': ...
45419                                                   []
Name: cast, Length: 45420, dtype: object

In [153]:
cast = []
for string in mc['cast']:
    cast_list = eval(str(string))
    if type(string)==float:
        cast.append(float('Nan'))
    else:
        cast2 = []
        for cast_dict in cast_list:
            cast_name = cast_dict["name"]
            cast2.append(cast_name)
        cast.append(cast2)

In [155]:
mc['cast'] = cast

In [156]:
mc['cast']

0        [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...
1        [Robin Williams, Jonathan Hyde, Kirsten Dunst,...
2        [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...
3        [Whitney Houston, Angela Bassett, Loretta Devi...
4        [Steve Martin, Diane Keaton, Martin Short, Kim...
                               ...                        
45415          [Leila Hatami, Kourosh Tahami, Elham Korda]
45416    [Angel Aquino, Perry Dizon, Hazel Orencio, Joe...
45417    [Erika Eleniak, Adam Baldwin, Julie du Page, J...
45418    [Iwan Mosschuchin, Nathalie Lissenko, Pavel Pa...
45419                                                   []
Name: cast, Length: 45420, dtype: object

Crew 전처리

In [128]:
mc[mc['crew']=='[]'].index

Int64Index([  189,   614,   635,   661,   711,   863,  1107,  1272,  1393,
             1405,
            ...
            45031, 45041, 45076, 45103, 45115, 45181, 45206, 45221, 45292,
            45315],
           dtype='int64', length=770)

In [157]:
type(mc['crew'][189])

str

In [146]:
mc['crew'] = mc['crew'].astype('str')

In [158]:
eval(mc['crew'][0])

[{'credit_id': '52fe4284c3a36847f8024f49',
  'department': 'Directing',
  'gender': 2,
  'id': 7879,
  'job': 'Director',
  'name': 'John Lasseter',
  'profile_path': '/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f4f',
  'department': 'Writing',
  'gender': 2,
  'id': 12891,
  'job': 'Screenplay',
  'name': 'Joss Whedon',
  'profile_path': '/dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f55',
  'department': 'Writing',
  'gender': 2,
  'id': 7,
  'job': 'Screenplay',
  'name': 'Andrew Stanton',
  'profile_path': '/pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f5b',
  'department': 'Writing',
  'gender': 2,
  'id': 12892,
  'job': 'Screenplay',
  'name': 'Joel Cohen',
  'profile_path': '/dAubAiZcvKFbboWlj7oXOkZnTSu.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f61',
  'department': 'Writing',
  'gender': 0,
  'id': 12893,
  'job': 'Screenplay',
  'name': 'Alec Sokolow',
  'profile_path': '/v79vlRYi94BZUQnkkyzn

In [159]:
# crew에서 name만 리스트로 추출하는 과정
crew1 = []
for string in mc['crew']:
    crew_list = eval(string)
    if type(string)==float:
        crew1.append(float('Nan'))
    else:
        crew2 = []
        for crew_dict in crew_list:
            crew_name = crew_dict["name"]
            crew2.append(crew_name)
        crew1.append(crew2)

In [160]:
# float 때문에 iteration이 되지 않아서 작업해줌
mc["cast"].loc[mc["cast"].isnull()] = mc["cast"].loc[mc["cast"].isnull()].apply(lambda x:[])
mc["crew"].loc[mc["crew"].isnull()] = mc["crew"].loc[mc["crew"].isnull()].apply(lambda x:[])
mc["genres"].loc[mc["genres"].isnull()] = mc["genres"].loc[mc["genres"].isnull()].apply(lambda x:[])

C:\Users\user\.conda\envs\test\lib\site-packages\pandas\core\indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [161]:
# name 중에서 director만 리스트로 추출하는 과정
crew = []
for i in range(len(crew1)):
    crew.append(' '.join(crew1[i][:1]))

In [162]:
mc['crew']= crew

In [163]:
# top3 actors를 추출하는 과정
actor= []
for i in range(len(mc['cast'])):
    actor.append(mc['cast'][i][:3])

In [164]:
# top3 genres를 추출하는 과정
genres= []
for i in range(len(mc['genres'])):
    genres.append(mc['genres'][i][:3])

In [165]:
mc['director'] = mc['crew']
mc['actor'] = actor
mc['genres'] = genres

In [167]:
mc[['title', 'actor', 'director', 'genres']].head(10)

title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   
5                         Heat   
6                      Sabrina   
7                 Tom and Huck   
8                 Sudden Death   
9                    GoldenEye   

                                               actor         director  \
0                [Tom Hanks, Tim Allen, Don Rickles]    John Lasseter   
1     [Robin Williams, Jonathan Hyde, Kirsten Dunst]  Larry J. Franco   
2         [Walter Matthau, Jack Lemmon, Ann-Margret]    Howard Deutch   
3  [Whitney Houston, Angela Bassett, Loretta Devine]  Forest Whitaker   
4         [Steve Martin, Diane Keaton, Martin Short]   Alan Silvestri   
5            [Al Pacino, Robert De Niro, Val Kilmer]     Michael Mann   
6        [Harrison Ford, Julia Ormond, Greg Kinnear]   Sydney Pollack   
7  [Jonathan Taylor Thomas, Brad Renfro, Rachael ...   David Loughery   
8  [Jean-Claude Van Damme, Powers Boothe, Dorian ...      Peter Hyams   
9     [Pierce Brosnan, Sean Bean, Izabella Scorupco]  Martin Campbell   

                          genres  
0    [Animation, Comedy, Family]  
1   [Adventure, Fantasy, Family]  
2              [Romance, Comedy]  
3       [Comedy, Drama, Romance]  
4                       [Comedy]  
5         [Action, Crime, Drama]  
6              [Comedy, Romance]  
7     [Action, Adventure, Drama]  
8  [Action, Adventure, Thriller]  
9  [Adventure, Action, Thriller]

이름을 구분하기 위해 처리하는 과정 (lowercase와 공백제거 활용)

In [168]:
def clean_director(x):
    return x.lower().replace(' ','')

In [169]:
mc['director'] = mc['director'].apply(lambda x:clean_director(x))

In [170]:
actor2 = []
for i in range(len(mc['actor'])):
    actor3 = []
    for j in range(len(mc['actor'][i])):
        actor3.append(mc['actor'][i][j].lower().replace(' ', ''))
    actor2.append(actor3)

In [171]:
mc['actor'] = actor2

In [172]:
genres2 = []
for i in range(len(mc['genres'])):
    genres3 = []
    for j in range(len(mc['genres'][i])):
        genres3.append(mc['genres'][i][j].lower())
    genres2.append(genres3)

In [173]:
mc['genres'] = genres2

In [174]:
mc[['title', 'actor', 'director', 'genres']].head(5)

title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            actor        director  \
0                [tomhanks, timallen, donrickles]    johnlasseter   
1     [robinwilliams, jonathanhyde, kirstendunst]   larryj.franco   
2        [waltermatthau, jacklemmon, ann-margret]    howarddeutch   
3  [whitneyhouston, angelabassett, lorettadevine]  forestwhitaker   
4         [stevemartin, dianekeaton, martinshort]   alansilvestri   

                         genres  
0   [animation, comedy, family]  
1  [adventure, fantasy, family]  
2             [romance, comedy]  
3      [comedy, drama, romance]  
4                      [comedy]

soup열 생성

In [175]:
def create_soup(x):
    return ' '.join(x['actor']) + ' ' + x['director'] + ' ' + ' '.join(x['genres'])

In [176]:
mc['soup'] = mc.apply(create_soup, axis=1)

CountVectorizer() 생성 

In [177]:
from sklearn.feature_extraction.text import CountVectorizer

In [178]:
cv = CountVectorizer(stop_words='english')
cv_matrix = cv.fit_transform(mc['soup'])

Similarity Score 계산

In [179]:
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
cosine_sim2 = cosine_similarity(cv_matrix, cv_matrix)

In [244]:
content_recommendations('The Dark Knight Rises', cosine_sim2)

10119          Batman Begins
12477        The Dark Knight
516        Romeo Is Bleeding
2989               Backdraft
4021                Hannibal
9227                  Shiner
23876              Quicksand
24817               Deadfall
9783         Amongst Friends
276      Murder in the First
dtype: object

In [245]:
content_recommendations('The Godfather', cosine_sim2)

1178             The Godfather: Part II
1913            The Godfather: Part III
15432                   The Rain People
1165                     Apocalypse Now
2004                      The Outsiders
3984                   Gardens of Stone
5260                        The Gambler
7924     The Night of the Following Day
18742                         Last Exit
34242                              Rege
dtype: object

앞에서 사용했던 content_recommendation 함수를 활용하면 앞서 사용했던 similarity score보다 더 정교한 추천 및 예측을 할 수 있다.